### ISSUE TITLE
[query] Add type-hinting functions

### ISSUE THREAD

**Issue by rnewman** [mozilla/mentat#474]
It would be convenient to have a way to constrain the types of variables explicitly:

```
[:find ?x :where
 [(uuid ?v)]
 [?x _ ?v]]
```

This hinting/constraining would be useful for queries like the above — find all entities associated with any UUID — but also allows for `ground` (and potentially other functions) to accept ambiguous EDN input. We don't have a good way right now to distinguish between `10` the `Long` and `10` the `entid`.

These functions are very straightforward to implement.

**Comment by rnewman** [comment]
Beginner's Guide®:

0. Make a new file, alongside `query-algebrizer/src/ground.rs`, called `hints.rs` that implements a function much like `apply_ground`; perhaps call it `apply_type_hint`. Leave the body of the function empty. Add the file to `query-algebrizer/src/lib.rs`, just as for `ground.rs`. Don't forget to `git add` it.
1. In `query-algebrizer/src/clauses/where_fn.rs`, you'll see a `match` block that dispatches on function name. Look at `core/src/lib.rs`'s definition of `ValueType` to see the type names for which we want to define hints. Add `match` arms to match.
2. Implement the body of `apply_type_hint`. It should take the type, get a variable argument, and call `self.constrain_var_to_type` on the variable. That function is defined in `query-algebrizer/src/clauses/mod.rs`.
3. Write tests, just like `ground`'s tests — you'll have some that test algebrizing, like `query-algebrizer/tests/ground.rs`, and some that work with real data, in the top-level `tests/query.rs`. You should be able to create a database, add two attributes of different types, add values for each attribute, and then run a query like:

```
[:find ?v
 :where
 [(instant ?v)]
 [_ _ ?v]]
```

to get only values of attributes that are `:db/valueType :db.type/instant`.

This should be equivalent to the query

```
[:find ?v
 :where
 [?a :db/valueType :db.type/instant]
 [_ ?a ?v]]
```

**Comment by rnewman** [comment]
Fixed in 98502eb6.
